lib/state/caching_ap.go

```diff
@@ -0,0 +1,200 @@
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	users, err := ap.GetUsers()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	userCAs, err := ap.GetCertAuthorities(services.UserCA, false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	hostCAs, err := ap.GetCertAuthorities(services.HostCA, false)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cs := &CachingAuthClient{
+		ap:         ap,
+		domainName: domainName,
+		nodes:      nodes,
+		proxies:    proxies,
+		users:      users,
+		userCAs:    userCAs,
+		hostCAs:    hostCAs,
+	}
+	return cs, nil
+}
+
+// GetDomainName is a part of auth.AccessPoint implementation
+func (cs *CachingAuthClient) GetDomainName() (string, error) {
+	cs.try(func() error {
+		dn, err := cs.ap.GetDomainName()
+		if err == nil {
+			cs.Lock()
+			defer cs.Unlock()
+			cs.domainName = dn
+		}
+		return err
+	})
+	return cs.domainName, nil
+}
+
+// GetNodes is a part of auth.AccessPoint implementation
+func (cs *CachingAuthClient) GetNodes() ([]services.Server, error) {
+	cs.try(func() error {
+		nodes, err := cs.ap.GetNodes()
+		if err == nil {
+			cs.Lock()
+			defer cs.Unlock()
+			cs.nodes = nodes
+		}
+		return err
+	})
+	return cs.nodes, nil
+}
+
+// GetProxies is a part of auth.AccessPoint implementation
+func (cs *CachingAuthClient) GetProxies() ([]services.Server, error) {
+	cs.try(func() error {
+		proxies, err := cs.ap.GetProxies()
+		if err == nil {
+			cs.Lock()
+			defer cs.Unlock()
+			cs.proxies = proxies
+		}
+		return err
+	})
+	return cs.proxies, nil
+}
+
+// GetCertAuthorities is a part of auth.AccessPoint implementation
+func (cs *CachingAuthClient) GetCertAuthorities(ct services.CertAuthType, loadKeys bool) ([]*services.CertAuthority, error) {
+	cs.try(func() error {
+		retval, err := cs.ap.GetCertAuthorities(ct, loadKeys)
+		if err == nil {
+			cs.Lock()
+			defer cs.Unlock()
+			if ct == services.UserCA {
+				cs.userCAs = retval
+			} else {
+				cs.hostCAs = retval
+			}
+		}
+		return err
+	})
+	if ct == services.UserCA {
+		return cs.userCAs, nil
+	}
+	return cs.hostCAs, nil
+}
+
+// GetUsers is a part of auth.AccessPoint implementation
+func (cs *CachingAuthClient) GetUsers() ([]services.User, error) {
+	cs.try(func() error {
+		users, err := cs.ap.GetUsers()
+		if err == nil {
+			cs.Lock()
+			defer cs.Unlock()
+			cs.users = users
+		}
+		return err
+	})
+	return cs.users, nil
+}
+
+// UpsertNode is part of auth.AccessPoint implementation
+func (cs *CachingAuthClient) UpsertNode(s services.Server, ttl time.Duration) error {
+	return cs.ap.UpsertNode(s, ttl)
+}
+
+// UpsertProxy is part of auth.AccessPoint implementation
+func (cs *CachingAuthClient) UpsertProxy(s services.Server, ttl time.Duration) error {
+	return cs.ap.UpsertProxy(s, ttl)
+}
+
+func (cs *CachingAuthClient) try(f func() error) {
+	if cs.lastErrorTime.Add(backoffDuration).After(time.Now()) {
+		// backoff: the recent error happend too recently
+		log.Error("BACKING OFF!!!! %v", f)
```

Comment:
can you make the comment a bit more descriptive. If I see one in the logs, it will scare me without giving to much context on why is it backing off and whether this is an error